ffi/derive/src/convert.rs

```diff
@@ -1,871 +1,497 @@
+            if is_fieldless_enum(&item) {
+                return derive_ffi_type_for_fieldless_enum(&input.ident, &item, &repr);
             }
-        }
-        syn::Data::Struct(item) => {
-            if is_transparent(input) {
-                derive_try_from_repr_c_for_transparent_item(&input.ident, &input.generics, item)
-            } else {
-                derive_try_from_repr_c_for_item(&input.ident)
-            }
-        }
-        syn::Data::Union(item) => abort!(item.union_token, "Unions are not supported"),
-    }
-}
-
-pub fn derive_into_ffi(input: &DeriveInput) -> TokenStream2 {
-    if !matches!(input.vis, syn::Visibility::Public(_)) {
-        abort!(input.vis, "Only public items are supported");
-    }
 
-    if is_opaque(input) {
-        if is_opaque_wrapper(input) {
-            return derive_into_ffi_for_opaque_item_wrapper(input);
+            derive_ffi_type_for_data_carrying_enum(&input.ident, input.generics, &item)
         }
-
-        return derive_into_ffi_for_opaque_item(input);
-    }
-
-    match &input.data {
-        syn::Data::Enum(item) => {
-            let repr = find_attr(&input.attrs, "repr");
-
-            if is_fieldless_enum(&input.ident, item, &repr) {
-                derive_into_ffi_for_fieldless_enum(&input.ident, &repr)
-            } else {
-                derive_into_ffi_for_item(&input.ident)
-            }
-        }
-        syn::Data::Struct(item) => {
-            if is_transparent(input) {
-                derive_into_ffi_for_transparent_item(&input.ident, &input.generics, item)
-            } else {
-                derive_into_ffi_for_item(&input.ident)
+        syn::Data::Struct(_) => {
+            if is_opaque(&input) {
+                return derive_ffi_type_for_opaque_struct(name, input.generics);
+            } else if is_opaque_wrapper(&input) {
+                return derive_ffi_type_for_extern_struct(name, input.generics);
             }
+
+            derive_ffi_type_for_repr_c_struct(name)
         }
-        syn::Data::Union(item) => abort!(item.union_token, "Unions are not supported"),
+        syn::Data::Union(item) => abort!(item.union_token, "Unions are not yet supported"),
     }
 }
 
-fn variant_discriminants(enum_: &syn::DataEnum) -> Vec<syn::Expr> {
-    let mut curr_discriminant: syn::Expr = parse_quote! {0};
-
-    enum_.variants.iter().fold(Vec::new(), |mut acc, variant| {
-        let discriminant = variant.discriminant.as_ref().map_or_else(
-            || curr_discriminant.clone(),
-            |discriminant| discriminant.1.clone(),
-        );
-
-        acc.push(discriminant.clone());
-        curr_discriminant = parse_quote! {
-            1 + #discriminant
-        };
-
-        acc
-    })
-}
-
-fn derive_try_from_repr_c_for_opaque_item_wrapper(input: &DeriveInput) -> TokenStream2 {
+fn derive_ffi_type_for_transparent_item(input: &syn::DeriveInput) -> TokenStream2 {
+    // TODO: We don't check to find which field is not a ZST.
```

Comment:
I'm not sure that this possible to do during macro expansion.

Reply:
I guess you are right. I suppose it could be controlled with derive macro helper attribute. Using the first field is good enough. ZSTs cannot be supported in FFI so the conversion should fail if the first  field is ZST